src/Processors/Formats/Impl/NativeORCBlockInputFormat.cpp

```diff
@@ -1285,52 +1286,48 @@ static ColumnPtr readOffsetsFromORCListColumn(const BatchType * orc_column)
 }
 
 static ColumnWithTypeAndName
-readColumnWithBooleanData(const orc::ColumnVectorBatch * orc_column, const orc::Type *, const String & column_name)
+readColumnWithBooleanData(const orc::ColumnVectorBatch * orc_column, const String & column_name)
 {
     const auto * orc_bool_column = dynamic_cast<const orc::LongVectorBatch *>(orc_column);
     auto internal_type = DataTypeFactory::instance().get("Bool");
     auto internal_column = internal_type->createColumn();
     auto & column_data = assert_cast<ColumnVector<UInt8> &>(*internal_column).getData();
-    column_data.reserve(orc_bool_column->numElements);
 
+    column_data.reserve(orc_bool_column->numElements);
     for (size_t i = 0; i < orc_bool_column->numElements; ++i)
-        column_data.push_back(static_cast<UInt8>(orc_bool_column->data[i]));
+    {
+        if (!orc_bool_column->hasNulls || orc_bool_column->notNull[i])
```

Comment:
@alexey-milovidov I'm curious why you modified the codes here? `readColumnWithBooleanData` and other similar functions don't need to consider processing `hasNull` or `notNull`. They are processed in the begining of `readColumnFromORCColumn`. I suspect that's what caused the performance degradation.

Reply:
Totally possible. But null map was read in the beginning of `readColumnFromORCColumn`, and then data column was read in `readColumnWithBooleanData`, and finally null map and data column was combined into `Nullable(Bool)`.